src/KX0231025IMU.cpp

```diff
@@ -19,82 +19,93 @@
-#define KX0231025_DATARATE_100HZ	 3
-#define KX0231025_DATARATE_200HZ	 4
-#define KX0231025_DATARATE_400HZ	 5
-#define KX0231025_DATARATE_800HZ	 6
-#define KX0231025_DATARATE_1600HZ	 7
-#define KX0231025_DATARATE_0_781HZ	 8
-#define KX0231025_DATARATE_1_563HZ	 9
-#define KX0231025_DATARATE_3_125HZ	 10
-#define KX0231025_DATARATE_6_25HZ	 11
-
-#define KX0231025_LOWPOWER_MODE		 0X00
-#define KX0231025_HIGHPOWER_MODE	 0X40
-
-
-KX0231025Class::KX0231025Class(TwoWire& wire, uint8_t slaveAddress) :
-	_wire(&wire),
-	_spi(NULL),
-	_slaveAddress(slaveAddress)
+#define KX0231025_ADDRESS 0x18
+
+#define KX0231025_WHO_AM_I_REG 0X0F
+#define KX0231025_CTRL1 0X18
+#define KX0231025_CTRL2 0X19
+#define KX0231025_CTRL3 0X20
+#define KX0231025_ODCNTL 0X1B
+
+#define KX0231025_STATUS_REG 0X15
+
+#define KX0231025_OUTX_L_XL 0X06
+#define KX0231025_OUTX_H_XL 0X07
+#define KX0231025_OUTY_L_XL 0X08
+#define KX0231025_OUTY_H_XL 0X09
+#define KX0231025_OUTZ_L_XL 0X10
+#define KX0231025_OUTZ_H_XL 0X11
+
+#define KX0231025_RANGE_2G 0
+#define KX0231025_RANGE_4G 1
+#define KX0231025_RANGE_8G 2
+
+#define KX0231025_DATARATE_12_5HZ 0
+#define KX0231025_DATARATE_25HZ 1
+#define KX0231025_DATARATE_50HZ 2
+#define KX0231025_DATARATE_100HZ 3
+#define KX0231025_DATARATE_200HZ 4
+#define KX0231025_DATARATE_400HZ 5
+#define KX0231025_DATARATE_800HZ 6
+#define KX0231025_DATARATE_1600HZ 7
+#define KX0231025_DATARATE_0_781HZ 8
+#define KX0231025_DATARATE_1_563HZ 9
+#define KX0231025_DATARATE_3_125HZ 10
+#define KX0231025_DATARATE_6_25HZ 11
+
+#define KX0231025_LOWPOWER_MODE 0X00
+#define KX0231025_HIGHPOWER_MODE 0X40
+
+KX0231025Class::KX0231025Class(TwoWire &wire, uint8_t slaveAddress) : _wire(&wire),
+																	  _spi(NULL),
+																	  _slaveAddress(slaveAddress),
+																	  _sda(0),
+																	  _scl(0)
 {
 }
-
-KX0231025Class::KX0231025Class(SPIClass& spi, int csPin) :
-	_wire(NULL),
-	_spi(&spi),
-	_csPin(csPin),
-	_spiSettings(10E6, MSBFIRST, SPI_MODE0)
+KX0231025Class::KX0231025Class(TwoWire &wire, uint8_t slaveAddress, int sda, int scl) : _wire(&wire),
+																						_spi(NULL),
+																						_slaveAddress(slaveAddress),
+																						_sda(sda),
+																						_scl(scl)
+{
+}
+KX0231025Class::KX0231025Class(SPIClass &spi, int csPin) : _wire(NULL),
+														   _spi(&spi),
+														   _csPin(csPin),
+														   _spiSettings(10E6, MSBFIRST, SPI_MODE0)
 {
 }
 
 KX0231025Class::~KX0231025Class()
 {
 }
-
 int KX0231025Class::begin(int powerMode, int accelerationRange, int outputDataRate)
 {
 	this->_accelerationRange = accelerationRange;
 
-	if (_spi != NULL) {
+	if (_spi != NULL)
+	{
 		pinMode(_csPin, OUTPUT);
 		digitalWrite(_csPin, HIGH);
 		_spi->begin();
 	}
-	else {
+	else if (_sda != 0)
+	{
+		_wire->begin(_sda, _scl);
+	}
+	else
+	{
 		_wire->begin();
 	}
 
-	if (readRegister(KX0231025_WHO_AM_I_REG) != 0x15) {
+	if (readRegister(KX0231025_WHO_AM_I_REG) != 0x15)
+	{
 		end();

```

Comment:
This doesnt look right. Should it be _wire->end()? Also maybe need a nullcheck here if spi is being used

Reply:
I didn't touch that code, except formatting the code. the end() function check if spi exist and choose what to close.

Reply:
Ah haha early in the day :)